src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+impl From<ErrorStack> for FromBytesError {
+    fn from(_: ErrorStack) -> Self {
+        Self::Implementation
+    }
+}
+
+//--- Formatting
+
+impl fmt::Display for FromBytesError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Self::UnsupportedAlgorithm => "algorithm not supported",
+            Self::InvalidKey => "malformed or insecure private key",
+            Self::Implementation => "an internal error occurred",
+        })
+    }
+}
+
+//--- Error
+
+impl std::error::Error for FromBytesError {}
+
+//----------- GenerateError --------------------------------------------------
+
+/// An error in generating a key pair with OpenSSL.
+#[derive(Clone, Debug)]
+pub enum GenerateError {
+    /// The requested algorithm was not supported.
+    UnsupportedAlgorithm,
+
+    /// An implementation failure occurred.
+    ///
+    /// This includes memory allocation failures.
+    Implementation,
+}
+
+//--- Conversion
+
+impl From<ErrorStack> for GenerateError {
+    fn from(_: ErrorStack) -> Self {
+        Self::Implementation
+    }
+}
+
+//--- Formatting
+
+impl fmt::Display for GenerateError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Self::UnsupportedAlgorithm => "algorithm not supported",
+            Self::Implementation => "an internal error occurred",
+        })
+    }
+}
+
+//--- Error
+
+impl std::error::Error for GenerateError {}
+
+//----------- DigestContext --------------------------------------------------
+
+/// Context for computing a message digest.
+pub struct DigestContext(Hasher);
+
+impl DigestContext {
+    /// Create a new context for a specified digest type.
+    pub fn new(digest_type: DigestType) -> Self {
+        Self(
+            match digest_type {
+                DigestType::Sha1 => Hasher::new(MessageDigest::sha1()),
+                DigestType::Sha256 => Hasher::new(MessageDigest::sha256()),
+                DigestType::Sha384 => Hasher::new(MessageDigest::sha384()),
+            }
+            .expect("assume that new cannot fail"),
+        )
+    }
+
+    /// Add input to the digest computation.
+    pub fn update(&mut self, data: &[u8]) {
+        self.0
+            .update(data)
+            .expect("assume that update does not fail")
+    }
+
+    /// Finish computing the digest.
+    pub fn finish(mut self) -> Digest {
+        Digest(self.0.finish().expect("assume that finish does not fail"))
+    }
+}
+
+//----------- Digest ---------------------------------------------------------
+
+/// A message digest.
+pub struct Digest(DigestBytes);
+
+impl AsRef<[u8]> for Digest {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+//----------- PublicKey ------------------------------------------------------
+
+/// A public key for verifying a signature.
+pub enum PublicKey {
+    /// Default variant, used for RSA.
+    Default(MessageDigest, PKey<Public>, u16),
```

Comment:
Suggest to rename this variant to `Rsa` as it is confusing reading code such as `verify()` and `dnskey()` when encountering the `Default` variant as it is not clear that it actually refers to RSA, as the comment even says on this variant.

Reply:
Done